{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import re\n",
    "import sys\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "#constants\n",
    "CYP2D6_PHENO=\"CYP2D6_Metabolyzer.csv\"\n",
    "CYP2C19_PHENO=\"CYP2C19_Metabolyzer.csv\"\n",
    "AIF=\"CustomArray_AIF.txt\"\n",
    "CYP2D6_AF=\"CYP2D6_AF.tsv\"\n",
    "RESOURCES_DIR=\"resources\"\n",
    "POP = 'European'\n",
    "SNPS = [\"rs489693\",\"rs4713916\",\"rs7997012\",\"rs6295\"]\n",
    "NTC_ID = 'NTC'\n",
    "\n",
    "cyp2d6_pheno_file = RESOURCES_DIR + \"/\" + CYP2D6_PHENO\n",
    "cyp2c19_pheno_file = RESOURCES_DIR + \"/\" + CYP2C19_PHENO\n",
    "cyp2d6_af_file = RESOURCES_DIR + \"/\" + CYP2D6_AF\n",
    "aif_file = RESOURCES_DIR + \"/\" + AIF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def now(sep=\":\"):\n",
    "    now = datetime.now()\n",
    "    current_time = now.strftime(\"%Y%M%D_%H{sep}%M{sep}%S\".format(sep=sep))\n",
    "    return current_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Diplotype</th>\n",
       "      <th>Activity_Score</th>\n",
       "      <th>Phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>*2/*41</td>\n",
       "      <td>0</td>\n",
       "      <td>EM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>*41/*2</td>\n",
       "      <td>0</td>\n",
       "      <td>EM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>*3/*3</td>\n",
       "      <td>0</td>\n",
       "      <td>PM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>*3/*3xN</td>\n",
       "      <td>0</td>\n",
       "      <td>PM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>*3/*4</td>\n",
       "      <td>0</td>\n",
       "      <td>PM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Diplotype Activity_Score Phenotype\n",
       "0    *2/*41              0        EM\n",
       "1    *41/*2              0        EM\n",
       "2     *3/*3              0        PM\n",
       "3   *3/*3xN              0        PM\n",
       "4     *3/*4              0        PM"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load metabolizer definitions\n",
    "CYP2D6_pheno = pd.read_csv(cyp2d6_pheno_file, sep=\"\\t\", header=0)\n",
    "CYP2C19_pheno = pd.read_csv(cyp2c19_pheno_file, sep=\"\\t\", header=0)\n",
    "CYP2D6_pheno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCBI SNP Reference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Assay ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>C__30634128_10</td>\n",
       "      <td>rs489693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>C____469857_10</td>\n",
       "      <td>rs4713916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>C__25986767_70</td>\n",
       "      <td>rs7997012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>C__27861809_10</td>\n",
       "      <td>rs6295</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               NCBI SNP Reference\n",
       "Assay ID                         \n",
       "C__30634128_10           rs489693\n",
       "C____469857_10          rs4713916\n",
       "C__25986767_70          rs7997012\n",
       "C__27861809_10             rs6295"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load array AIF file\n",
    "AIF_table = pd.read_csv(aif_file, sep=\"\\t\", header=0, comment='*', encoding = \"ISO-8859-1\", usecols=['Assay ID','NCBI SNP Reference'], index_col='Assay ID')\n",
    "AIF_table.loc[list(geno_table.columns)[1:5],'NCBI SNP Reference'] = SNPS\n",
    "AIF_table.loc[list(geno_table.columns)[1:5]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>African</th>\n",
       "      <th>African_American</th>\n",
       "      <th>European</th>\n",
       "      <th>Near_Eastern</th>\n",
       "      <th>East_Asian</th>\n",
       "      <th>South_Asian</th>\n",
       "      <th>American</th>\n",
       "      <th>Latino</th>\n",
       "      <th>Oceanian</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>*1</td>\n",
       "      <td>0.3423</td>\n",
       "      <td>0.3482</td>\n",
       "      <td>0.4138</td>\n",
       "      <td>0.3803</td>\n",
       "      <td>0.3489</td>\n",
       "      <td>0.4854</td>\n",
       "      <td>0.5883</td>\n",
       "      <td>0.4620</td>\n",
       "      <td>0.7300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>*2</td>\n",
       "      <td>0.1902</td>\n",
       "      <td>0.1428</td>\n",
       "      <td>0.2517</td>\n",
       "      <td>0.2184</td>\n",
       "      <td>0.1332</td>\n",
       "      <td>0.2910</td>\n",
       "      <td>0.2228</td>\n",
       "      <td>0.2543</td>\n",
       "      <td>0.0120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>*4</td>\n",
       "      <td>0.0348</td>\n",
       "      <td>0.0640</td>\n",
       "      <td>0.1793</td>\n",
       "      <td>0.1174</td>\n",
       "      <td>0.0079</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>0.1086</td>\n",
       "      <td>0.1212</td>\n",
       "      <td>0.0248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>*41</td>\n",
       "      <td>0.1241</td>\n",
       "      <td>0.0637</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>0.1764</td>\n",
       "      <td>0.0234</td>\n",
       "      <td>0.1295</td>\n",
       "      <td>0.0253</td>\n",
       "      <td>0.0442</td>\n",
       "      <td>0.0088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>*35</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0077</td>\n",
       "      <td>0.0519</td>\n",
       "      <td>0.0230</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>0.0110</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0.0239</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        African  African_American  European  Near_Eastern  East_Asian  \\\n",
       "Allele                                                                  \n",
       "*1       0.3423            0.3482    0.4138        0.3803      0.3489   \n",
       "*2       0.1902            0.1428    0.2517        0.2184      0.1332   \n",
       "*4       0.0348            0.0640    0.1793        0.1174      0.0079   \n",
       "*41      0.1241            0.0637    0.0888        0.1764      0.0234   \n",
       "*35      0.0000            0.0077    0.0519        0.0230      0.0010   \n",
       "\n",
       "        South_Asian  American  Latino  Oceanian  \n",
       "Allele                                           \n",
       "*1           0.4854    0.5883  0.4620    0.7300  \n",
       "*2           0.2910    0.2228  0.2543    0.0120  \n",
       "*4           0.0888    0.1086  0.1212    0.0248  \n",
       "*41          0.1295    0.0253  0.0442    0.0088  \n",
       "*35          0.0110    0.0075  0.0239    0.0000  "
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load CYP2D6 alleles frequencies \n",
    "cyp2d6_af_table = pd.read_csv(cyp2d6_af_file, sep=\"\\t\", header=0, index_col='Allele')\n",
    "cyp2d6_af_table.fillna(0, inplace=True)\n",
    "cyp2d6_af_table.sort_values(POP, ascending=False, inplace=True)\n",
    "cyp2d6_af_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinician</th>\n",
       "      <th>Group_PGx</th>\n",
       "      <th>first_drug</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>RF16_1</td>\n",
       "      <td>B01</td>\n",
       "      <td>1</td>\n",
       "      <td>Amisulpride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_2</td>\n",
       "      <td>B01</td>\n",
       "      <td>1</td>\n",
       "      <td>Venlafaxina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_3</td>\n",
       "      <td>B01</td>\n",
       "      <td>0</td>\n",
       "      <td>Escitalopram</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_4</td>\n",
       "      <td>T01</td>\n",
       "      <td>0</td>\n",
       "      <td>Escitalopram</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_5</td>\n",
       "      <td>T01</td>\n",
       "      <td>1</td>\n",
       "      <td>Amisulpride</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Clinician  Group_PGx    first_drug\n",
       "Sample                                   \n",
       "RF16_1       B01          1   Amisulpride\n",
       "RF16_2       B01          1   Venlafaxina\n",
       "RF16_3       B01          0  Escitalopram\n",
       "RF16_4       T01          0  Escitalopram\n",
       "RF16_5       T01          1   Amisulpride"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Sample details\n",
    "sample_cols = [c-1 for c in [1,2,3,4]]\n",
    "samples_info = pd.read_csv(\"samples_info.csv\", sep=\"\\t\", header=0, usecols=sample_cols)\n",
    "samples_info.columns=['Clinician','Sample','Group_PGx','first_drug']\n",
    "samples_info.set_index('Sample',inplace=True)\n",
    "samples_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CYP2D6</th>\n",
       "      <th>CYP2C19</th>\n",
       "      <th>CYP2D6_pheno</th>\n",
       "      <th>CYP2C19_pheno</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>RF16_1</td>\n",
       "      <td>*1/*1x2</td>\n",
       "      <td>*1/*2</td>\n",
       "      <td>UM</td>\n",
       "      <td>IM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_2</td>\n",
       "      <td>*4/*9</td>\n",
       "      <td>*1/*2</td>\n",
       "      <td>IM</td>\n",
       "      <td>IM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_3</td>\n",
       "      <td>*34/*39</td>\n",
       "      <td>*1/*1</td>\n",
       "      <td>EM</td>\n",
       "      <td>IM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_4</td>\n",
       "      <td>*34/*39</td>\n",
       "      <td>*1/*1</td>\n",
       "      <td>EM</td>\n",
       "      <td>IM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_5</td>\n",
       "      <td>*4/*69</td>\n",
       "      <td>*1/*17</td>\n",
       "      <td>PM</td>\n",
       "      <td>EM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         CYP2D6 CYP2C19 CYP2D6_pheno CYP2C19_pheno\n",
       "Sample                                            \n",
       "RF16_1  *1/*1x2   *1/*2          UM             IM\n",
       "RF16_2    *4/*9   *1/*2          IM             IM\n",
       "RF16_3  *34/*39   *1/*1          EM             IM\n",
       "RF16_4  *34/*39   *1/*1          EM             IM\n",
       "RF16_5   *4/*69  *1/*17           PM            EM"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CYP alleles from from allele typer\n",
    "CYP_alleles_table = pd.read_csv(\"detail_result.txt\", sep=\"\\t\", skiprows=10, encoding = \"ISO-8859-1\", usecols=['sample ID','CYP2D6','CYP2C19'])\n",
    "CYP_alleles_table.rename(columns={'sample ID' : 'Sample'}, inplace=True)\n",
    "CYP_alleles_table.set_index('Sample', inplace=True)\n",
    "indexCYP = CYP_alleles_table.index.values\n",
    "indexCYP[0:9] = list(samples_info.index)\n",
    "col_idx = CYP_alleles_table.columns.get_loc('CYP2D6')\n",
    "\n",
    "if NTC_ID in CYP_alleles_table.index.values:\n",
    "    CYP_alleles_table.drop(NTC_ID, inplace=True)\n",
    "if 'Run date' in CYP_alleles_table.index.values:\n",
    "    CYP_alleles_table.drop('Run date', inplace=True)\n",
    "    \n",
    "for idx, value in enumerate(CYP_alleles_table['CYP2D6'].values):\n",
    "    genos = value.split(', ')\n",
    "    af_mean = []\n",
    "    for g in genos:\n",
    "        g = re.sub(r'x[0-9]+','xN', g)\n",
    "        alleles = g.split('/')\n",
    "        af_sum = sum(a for a in cyp2d6_af_table.loc[alleles,POP])\n",
    "        af_mean.append(af_sum / 2)\n",
    "    \n",
    "    sort_idx = sorted(range(len(af_mean)), key=lambda k: af_mean[k])\n",
    "    CYP_alleles_table.iloc[idx,col_idx] = genos[sort_idx[0]]\n",
    "\n",
    "CYP_alleles_table.head()\n",
    "\n",
    "CYP_alleles_table['CYP2D6_pheno'] = \"\"\n",
    "CYP_alleles_table['CYP2C19_pheno'] = \"\"\n",
    "\n",
    "cnv = re.compile('(\\*[0-9]+)x([0-9]+)')\n",
    "col_idx = CYP_alleles_table.columns.get_loc('CYP2D6_pheno')\n",
    "for idx, geno in enumerate(CYP_alleles_table['CYP2D6'].values):\n",
    "    m = cnv.search(geno)\n",
    "    if m is not None:\n",
    "        if m.group(2) == \"2\":\n",
    "            if geno in CYP2D6_pheno['Diplotype'].values:\n",
    "                CYP_alleles_table.iloc[idx,col_idx]=CYP2D6_pheno.loc[CYP2D6_pheno['Diplotype']==geno, 'Phenotype'].values[0]\n",
    "            else:\n",
    "                geno_new = re.sub(r'x[3-9]+','xN', geno)\n",
    "                if geno_new in CYP2D6_pheno['Diplotype'].values:\n",
    "                    CYP_alleles_table.iloc[idx,col_idx]=CYP2D6_pheno.loc[CYP2D6_pheno['Diplotype']==geno_new, 'Phenotype'].values[0]\n",
    "                else:\n",
    "                    CYP_alleles_table.iloc[idx,col_idx]=\"UNDETERMINED\"\n",
    "        elif m.group(2) > 2:\n",
    "            geno_new = re.sub(r'x[3-9]+','xM', geno)\n",
    "            if geno_new in CYP2D6_pheno['Diplotype'].values:\n",
    "                CYP_alleles_table.iloc[idx,col_idx]=CYP2D6_pheno.loc[CYP2D6_pheno['Diplotype']==geno_new, 'Phenotype'].values[0]\n",
    "            else:\n",
    "                geno_new = re.sub(r'x[3-9]+','xN', geno)\n",
    "                if geno_new in CYP2D6_pheno['Diplotype'].values:\n",
    "                    CYP_alleles_table.iloc[idx,col_idx]=CYP2D6_pheno.loc[CYP2D6_pheno['Diplotype']==geno_new, 'Phenotype'].values[0]\n",
    "                else:\n",
    "                    CYP_alleles_table.iloc[idx,col_idx]=\"UNDETERMINED\"\n",
    "    else:\n",
    "        if geno in CYP2D6_pheno['Diplotype'].values:\n",
    "            CYP_alleles_table.iloc[idx,col_idx]=CYP2D6_pheno.loc[CYP2D6_pheno['Diplotype']==geno, 'Phenotype'].values[0]\n",
    "        else:\n",
    "            CYP_alleles_table.iloc[idx,col_idx]=\"UNDETERMINED\"\n",
    "\n",
    "col_idx = CYP_alleles_table.columns.get_loc('CYP2C19_pheno')\n",
    "for idx, geno in enumerate(CYP_alleles_table['CYP2C19'].values):\n",
    "    if geno in CYP2C19_pheno['Diplotype'].values:\n",
    "        CYP_alleles_table.iloc[idx,col_idx]=CYP2C19_pheno.loc[CYP2C19_pheno['Diplotype']==geno, 'Phenotype'].values[0]\n",
    "    else:\n",
    "        CYP_alleles_table.iloc[idx,col_idx]=\"UNDETERMINED\"\n",
    "CYP_alleles_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "if ($PM>=2 && $UM==0 && $IM==0 && $EM==0) {\n",
    "    $output{$line[0]}{CYP2C19}{pheno}=\"PM\";\n",
    "} elsif ((($IM==1 || $IM==2) && $PM <= 1 && $EM==0) || ($EM==1 && $PM <=1)) {\n",
    "    $output{$line[0]}{CYP2C19}{pheno}=\"IM\";\n",
    "} elsif ($EM==1 && $PM==0 && $IM==1) {\n",
    "    $output{$line[0]}{CYP2C19}{pheno}=\"EM\";\n",
    "} elsif ($EM==2 || (($EM + $IM) > 2)) {\n",
    "    $output{$line[0]}{CYP2C19}{pheno}=\"UM\";\n",
    "} else {$output{$line[0]}{CYP2C19}{pheno}=\"NA\"}\n",
    "\n",
    "$output{$line[0]}{CYP2C19}{alleles}=$mystring[0];\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genotyping results from TaqMan Genotyper\n",
    "geno_table = pd.read_csv(\"Genotype Matrix.csv\", sep=\",\", skiprows=15, encoding = \"ISO-8859-1\")\n",
    "geno_table.rename(columns=AIF_table.to_dict()['NCBI SNP Reference'], inplace=True)\n",
    "geno_table = geno_table[['Sample/Assay'] + SNPS]\n",
    "geno_table.rename(columns={'Sample/Assay':'Sample'}, inplace=True)\n",
    "geno_table.set_index('Sample', inplace=True)\n",
    "indexNamesArr = geno_table.index.values\n",
    "indexNamesArr[0:9] = list(samples_info.index)\n",
    "#geno_table.reindex(labels=list(samples_info.index))\n",
    "geno_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NA04671' 'NA17004' 'NA17005' 'NA17034' 'NA17051' 'NA17053' 'NA17055'\n",
      " 'NA17057' 'NA17059' 'NA17060' 'NA17104' 'NA17105' 'NA17108' 'NA17109'\n",
      " 'NA17125' 'NA17201' 'NA17202' 'NA17203' 'NA17205' 'NA17208' 'NTC']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SLC6A4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>RF16_1</td>\n",
       "      <td>LL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_2</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_3</td>\n",
       "      <td>LL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_4</td>\n",
       "      <td>LS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_5</td>\n",
       "      <td>LS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SLC6A4\n",
       "Sample       \n",
       "RF16_1     LL\n",
       "RF16_2     SS\n",
       "RF16_3     LL\n",
       "RF16_4     LS\n",
       "RF16_5     LS"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SLC6A4 long-short alleles\n",
    "SLC6A4_table = pd.read_csv(\"example_input_files/SLC6A4_LS_example.tsv\", sep=\"\\t\", header=0, index_col='Sample')\n",
    "indexSLC = SLC6A4_table.index.values\n",
    "indexSLC[0:9] = list(samples_info.index)\n",
    "SLC6A4_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotype table contains samples not found in sample information table\n",
      "Allele-typer table contains samples not found in sample information table\n",
      "SLC6A4 table contains samples not found in sample inforamtion table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinician</th>\n",
       "      <th>Group_PGx</th>\n",
       "      <th>first_drug</th>\n",
       "      <th>CYP2D6</th>\n",
       "      <th>CYP2C19</th>\n",
       "      <th>CYP2D6_pheno</th>\n",
       "      <th>CYP2C19_pheno</th>\n",
       "      <th>SLC6A4</th>\n",
       "      <th>rs489693</th>\n",
       "      <th>rs4713916</th>\n",
       "      <th>rs7997012</th>\n",
       "      <th>rs6295</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>RF16_1</td>\n",
       "      <td>B01</td>\n",
       "      <td>1</td>\n",
       "      <td>Amisulpride</td>\n",
       "      <td>*1/*1x2</td>\n",
       "      <td>*1/*2</td>\n",
       "      <td>UM</td>\n",
       "      <td>IM</td>\n",
       "      <td>LL</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/T</td>\n",
       "      <td>A/G</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_2</td>\n",
       "      <td>B01</td>\n",
       "      <td>1</td>\n",
       "      <td>Venlafaxina</td>\n",
       "      <td>*4/*9</td>\n",
       "      <td>*1/*2</td>\n",
       "      <td>IM</td>\n",
       "      <td>IM</td>\n",
       "      <td>SS</td>\n",
       "      <td>C/C</td>\n",
       "      <td>NOAMP</td>\n",
       "      <td>NOAMP</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_3</td>\n",
       "      <td>B01</td>\n",
       "      <td>0</td>\n",
       "      <td>Escitalopram</td>\n",
       "      <td>*34/*39</td>\n",
       "      <td>*1/*1</td>\n",
       "      <td>EM</td>\n",
       "      <td>IM</td>\n",
       "      <td>LL</td>\n",
       "      <td>NOAMP</td>\n",
       "      <td>NOAMP</td>\n",
       "      <td>NOAMP</td>\n",
       "      <td>NOAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_4</td>\n",
       "      <td>T01</td>\n",
       "      <td>0</td>\n",
       "      <td>Escitalopram</td>\n",
       "      <td>*34/*39</td>\n",
       "      <td>*1/*1</td>\n",
       "      <td>EM</td>\n",
       "      <td>IM</td>\n",
       "      <td>LS</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RF16_5</td>\n",
       "      <td>T01</td>\n",
       "      <td>1</td>\n",
       "      <td>Amisulpride</td>\n",
       "      <td>*4/*69</td>\n",
       "      <td>*1/*17</td>\n",
       "      <td>PM</td>\n",
       "      <td>EM</td>\n",
       "      <td>LS</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/T</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Clinician  Group_PGx    first_drug   CYP2D6 CYP2C19 CYP2D6_pheno  \\\n",
       "Sample                                                                    \n",
       "RF16_1       B01          1   Amisulpride  *1/*1x2   *1/*2          UM    \n",
       "RF16_2       B01          1   Venlafaxina    *4/*9   *1/*2          IM    \n",
       "RF16_3       B01          0  Escitalopram  *34/*39   *1/*1          EM    \n",
       "RF16_4       T01          0  Escitalopram  *34/*39   *1/*1          EM    \n",
       "RF16_5       T01          1   Amisulpride   *4/*69  *1/*17           PM   \n",
       "\n",
       "       CYP2C19_pheno SLC6A4 rs489693 rs4713916 rs7997012 rs6295  \n",
       "Sample                                                           \n",
       "RF16_1            IM     LL      C/C       C/T       A/G    G/G  \n",
       "RF16_2            IM     SS      C/C     NOAMP     NOAMP    G/G  \n",
       "RF16_3            IM     LL    NOAMP     NOAMP     NOAMP  NOAMP  \n",
       "RF16_4            IM     LS      C/C       C/C       G/G    G/G  \n",
       "RF16_5            EM     LS      C/C       C/T       G/G    G/G  "
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Check samples in genotype matrix and CYP alleles are included in sample info\n",
    "result =  all(elem in samples_info.index.values for elem in geno_table.index.values)\n",
    "if not result:\n",
    "    print(\"Genotype table contains samples not found in sample information table\")\n",
    "    \n",
    "result =  all(elem in samples_info.index.values for elem in CYP_alleles_table.index.values)\n",
    "if not result:\n",
    "    print(\"Allele-typer table contains samples not found in sample information table\")\n",
    "\n",
    "result =  all(elem in samples_info.index.values for elem in SLC6A4_table.index.values)\n",
    "if not result:\n",
    "    print(\"SLC6A4 table contains samples not found in sample inforamtion table\")\n",
    "    \n",
    "#Merge sample info + genotypes + SLC6A4 alleles + CYP alleles \n",
    "samples_info = samples_info.merge(CYP_alleles_table, on='Sample')\n",
    "samples_info = samples_info.merge(SLC6A4_table, on='Sample')\n",
    "samples_info = samples_info.merge(geno_table, on='Sample')\n",
    "samples_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_info.to_csv(\"test.tsv\",sep=\"\\t\", header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
